lib/AST/TypeCheckRequests.cpp

```diff
@@ -1742,6 +1742,19 @@ ArgumentList *UnresolvedMacroReference::getArgs() const {
   llvm_unreachable("Unhandled case");
 }
 
+MacroRoles UnresolvedMacroReference::getMacroRoles() const {
+  if (pointer.is<MacroExpansionExpr *>())
+    return MacroRole::Expression;
+
+  if (pointer.is<MacroExpansionDecl *>())
+    return MacroRole::Declaration;
```

Comment:
I'm don't think this computation is always accurate, because the parser can't tell whether a freestanding macro in code item position will be expression or declaration.

Reply:
Ah, I don't think we need to distinguish between those cases for macro resolution. I'll try using `getFreestandingMacroRoles()` instead

Reply:
Agreed. There could be cases where two freestanding macros (one `codeItem` and one `declaration`) sharing the same name and signature, and type contexts are the only place where a reference to such a macro is not ambiguous. But overloading macro declarations this way sounds like a A Bad Idea™. Such redeclarations could also be diagnosed earlier.